### PR TITLE
✨ MHC: implement annotation to manually mark machines for remediation

### LINKED ISSUE #10197: Manual remediation of Machines
### What would you like to be added (User Story)?

As an operator, I would like to be able to recreate single machines in a controlled way

### Detailed Description

The traditional way to re-create a single Machine is to delete it manually.

However, directly deleting a machine has some well-known downsides:
- Once the operation is done, it cannot be reverted
- The operation is "low level", it skips several system safeguards and this can be risky especially when you delete control plane nodes (see discussion in https://github.com/kubernetes-sigs/cluster-api/issues/9919). More speifically
  - With manual deletion, all the checks about [max unhealthy](https://github.com/kubernetes-sigs/cluster-api/blob/0f47a19e038ee6b0d3b1e7675a62cdaf84face8c/api/v1beta1/machinehealthcheck_types.go#L53-L65), are implicitly skipped
  - With manual deletion, all the checks about the fact the machine can be [safely remediated](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20191017-kubeadm-based-control-plane.md#remediation-using-delete-and-recreate), are implicitly skipped
  
This proposal is about introducing a new annotation (name TBD) to be applied to a machine that the operator wants to safely delete (and thus to leave "low level" deletion only as the ultimate escape patch).

This annotation will be processed by the MHC controller in [healthCheckTargets](https://github.com/kubernetes-sigs/cluster-api/blob/0f47a19e038ee6b0d3b1e7675a62cdaf84face8c/internal/controllers/machinehealthcheck/machinehealthcheck_targets.go#L283), and this the machine will be included in the pool of machines to be considered for remediation.
 
MHC, will then determine if there are the conditions to remediate such a machine, and the remediation with then be taken charge of by the owning controller or by the external remediation tool.

The remediation owner will then perform another round of checks, and finally transform the manual remediation in an actual machine deletion.
 
 

### Anything else you would like to add?

Once this is in place, we can eventually think of strategies to give higher priority for deletion to machines being remediated/manually remediated, in case more machines are being deleted at the same time.
However this is less important than providing a safe path to deletion (and thus IMO not blocking for a first iteration)

### Label(s) to be applied

/kind feature
/area machinehealthcheck
/triage accepted

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Implements annotation `cluster.x-k8s.io/remediate-machine` which could get manually set on machines to mark machines for remediation via a MHC.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10197

/area machinehealthcheck

### CODE REVIEW

File: api/v1beta1/common_types.go [hunk 1]
```diff
@@ -128,6 +128,9 @@ const (
 	// MachineSkipRemediationAnnotation is the annotation used to mark the machines that should not be considered for remediation by MachineHealthCheck reconciler.
 	MachineSkipRemediationAnnotation = "cluster.x-k8s.io/skip-remediation"
 
+	// RemediateMachineAnnotation is the annotation used to mark machines that should be remediated by MachineHealthCheck reconciler.
+	RemediateMachineAnnotation = "cluster.x-k8s.io/remediate-machine"
```
Comment (chrischdi) [inline api/v1beta1/common_types.go:132]: Naming to be finalized

File: api/v1beta1/condition_consts.go [hunk 2]
```diff
@@ -152,6 +152,9 @@ const (
 	// MachineHasFailureReason is the reason used when a machine has either a FailureReason or a FailureMessage set on its status.
 	MachineHasFailureReason = "MachineHasFailure"
 
+	// RemediateMachineReason is the reason used when a machine has either a ManualRemediationReason or a ManualRemediationMessage set on its status.
+	RemediateMachineReason = "MachineManualRemediation"
```
Comment (chrischdi) [inline api/v1beta1/condition_consts.go:156]: Naming to be finalized

File: api/v1beta1/common_types.go (hunk 1 above)
Comment (fabriziopandini) [reply api/v1beta1/common_types.go:132]: What about `cluster.x-k8s.io/force-remediation` (to express the intent) or `cluster.x-k8s.io/manual-remediation` (similar, not sure if equally clear)

Comment (sbueringer) [reply api/v1beta1/common_types.go:132]: I thought a bit about it. I think I like the current annotation the most:

* It's consistent with `cluster.x-k8s.io/delete-machine`
* `cluster.x-k8s.io/force-remediation` for me implies that we force a remediation in a sense that we skip something that would be usually done. Given that the idea here is not skip the usual MHC safeguards I think "force" might be misleading to some users.
* `cluster.x-k8s.io/manual-remediation` I think this unnecessarily limits the semantics of the annotation. We don't actually care if a users sets this annotation manually or e.g. if some other health check system detects an issue and then sets the annotation (without any manual user action)

File: api/v1beta1/condition_consts.go (hunk 2 above)
Comment (sbueringer) [reply api/v1beta1/condition_consts.go:156]: The value should be consistent with the name of the constant

Comment (sbueringer) [reply api/v1beta1/condition_consts.go:156]: Similar to what I wrote on the annotation. I would keep this more generic. So I would use e.g. `HasRemediateMachineAnnotation`. We should not assume that someone set this annotation manually.
(same for the labels_and_annotations.md and some other places in this PR)

The description is very confusing to me, maybe outdated?

File: internal/controllers/machinehealthcheck/machinehealthcheck_targets.go [hunk 3]
```diff
@@ -82,6 +82,7 @@ func (t *healthCheckTarget) nodeName() string {
 
 // Determine whether or not a given target needs remediation.
 // The node will need remediation if any of the following are true:
+// - The Machine has the manual remediation annotation
```
Comment (sbueringer) [inline internal/controllers/machinehealthcheck/machinehealthcheck_targets.go:85]:
```suggestion
// - The Machine has the remediate machine annotation
```

File: internal/controllers/machinehealthcheck/machinehealthcheck_targets.go [hunk 4]
```diff
@@ -93,6 +94,12 @@ func (t *healthCheckTarget) needsRemediation(logger logr.Logger, timeoutForMachi
 	var nextCheckTimes []time.Duration
 	now := time.Now()
 
+	if annotations.HasRemediateMachine(t.Machine) {
+		conditions.MarkFalse(t.Machine, clusterv1.MachineHealthCheckSucceededCondition, clusterv1.RemediateMachineReason, clusterv1.ConditionSeverityWarning, "Marked for manual remediation")
```
Comment (sbueringer) [inline internal/controllers/machinehealthcheck/machinehealthcheck_targets.go:98]:
```suggestion
		conditions.MarkFalse(t.Machine, clusterv1.MachineHealthCheckSucceededCondition, clusterv1.RemediateMachineReason, clusterv1.ConditionSeverityWarning, "Marked for remediation via remediate-machine annotation")
```
(similar in log below)

File: internal/controllers/machinehealthcheck/machinehealthcheck_targets_test.go [hunk 5]
```diff
@@ -340,6 +340,18 @@ func TestHealthCheckTargets(t *testing.T) {
 	}
 	machineFailureMsgCondition := newFailedHealthCheckCondition(clusterv1.MachineHasFailureReason, "FailureMessage: %s", failureMsg)
 
+	// Target for when the machine has a failure message
```
Comment (sbueringer) [inline internal/controllers/machinehealthcheck/machinehealthcheck_targets_test.go:343]: copy & paste?

File: util/annotations/helpers.go [hunk 6]
```diff
@@ -48,6 +48,11 @@ func HasSkipRemediation(o metav1.Object) bool {
 	return hasAnnotation(o, clusterv1.MachineSkipRemediationAnnotation)
 }
 
+// HasRemediateMachine returns true if the object has the `skip-remediation` annotation.
```
Comment (sbueringer) [inline util/annotations/helpers.go:51]: copy & paste

File: api/v1beta1/condition_consts.go (hunk 2 above)
Comment (chrischdi) [reply api/v1beta1/condition_consts.go]: Adjusted to

`HasRemediateMachineAnnotationReason = "HasRemediateMachineAnnotation"`

File: api/v1beta1/condition_consts.go [hunk 7]
```diff
@@ -152,6 +152,9 @@ const (
 	// MachineHasFailureReason is the reason used when a machine has either a FailureReason or a FailureMessage set on its status.
 	MachineHasFailureReason = "MachineHasFailure"
 
+	// HasRemediateMachineAnnotationReason is the reason used when a machine has either a ManualRemediationReason or a ManualRemediationMessage set on its status.
```
Comment (sbueringer) [inline api/v1beta1/condition_consts.go:155]: I think this description is outdated / copy&paste (? :))

**Review by sbueringer — COMMENTED** [review]
Thx, last nit

**Comment by sbueringer** [comment]
/lgtm

Feel free to squash

**Pull request merged**